File: src/main/java/ml/wonwoo/springdatashowcase/order/OrderEventListener.java
package ml.wonwoo.springdatashowcase.order;

import lombok.extern.slf4j.Slf4j;
import ml.wonwoo.springdatashowcase.event.Event;
import ml.wonwoo.springdatashowcase.event.EventRepository;
import ml.wonwoo.springdatashowcase.order.Order.OrderPaid;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;
import org.springframework.util.Assert;

@Component
@Slf4j
public class OrderEventListener {

    private final EventRepository eventRepository;

    public OrderEventListener(EventRepository eventRepository) {

        Assert.notNull(eventRepository, "eventRepository must not be null!");

        this.eventRepository = eventRepository;

    }

    @EventListener
    public void orderPaid(OrderPaid orderPaid) {

        Order order = orderPaid.getOrder();

        eventRepository.save(new Event(order));

        log.info("Publisher event {}", order);

    }
}

File: kotlin/src/main/kotlin/ml/wonwoo/springdatashowcase/order/OrderEventListener.kt
package ml.wonwoo.springdatashowcase.order

import ml.wonwoo.springdatashowcase.event.Event
import ml.wonwoo.springdatashowcase.event.EventRepository
import ml.wonwoo.springdatashowcase.order.Order.*
import ml.wonwoo.springdatashowcase.support.logger
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component

@Component
class OrderEventListener(private val eventRepository: EventRepository) {

    companion object {
        val logger = logger<OrderEventListener>()
    }

    @EventListener
    fun orderPaid(orderPaid: OrderPaid) {

        val order = orderPaid.order

        this.eventRepository.save(Event(order))

        logger.info("Publisher event $order")

    }
}